cabal-install-solver/src/Distribution/Solver/Modular/Message.hs

```diff
@@ -206,12 +208,81 @@ data MergedPackageConflict = MergedPackageConflict {
   , versionConflict :: Maybe VR
   }
 
-showQPNPOpt :: QPN -> POption -> String
-showQPNPOpt qpn@(Q _pp pn) (POption i linkedTo) =
+data ProgressAction =
+    Trying
+  | Skipping
+  | Rejecting
+
+instance Show ProgressAction where
+  show Trying = "trying: "
+  show Skipping = "skipping: "
+  show Rejecting = "rejecting: "
+
+showQFNBool :: ProgressAction -> QFN -> Bool -> String
+showQFNBool a q b = show a ++ Flag.showQFNBool q b
+
+showQSNBool :: ProgressAction -> QSN -> Bool -> String
+showQSNBool a q b = show a ++ Flag.showQSNBool q b
+
+showOptions :: ProgressAction -> QPN -> [POption] -> String
+showOptions a q [p] = show a ++ showOption q p
+showOptions a q ps = show a ++ showIsOrVs q (tryVs ps)
+
+showOption :: QPN -> POption -> String
+showOption qpn@(Q _pp pn) (POption i linkedTo) =
   case linkedTo of
     Nothing  -> showPI (PI qpn i) -- Consistent with prior to POption
     Just pp' -> showQPN qpn ++ "~>" ++ showPI (PI (Q pp' pn) i)
 
+-- | A list of versions, or a list of instances.
+data IsOrVs = Is [POption] | Vs [Ver] deriving Show
+
+-- | Try to convert a list of options to a list of versions, or a list of
+-- instances if any of the options is linked (installed).  Singleton lists or
+-- empty lists are always converted to Is.
+-- >>> tryVs [v0, v1]
+-- Vs [mkVersion [0],mkVersion [1]]
+-- >>> tryVs [v0]
+-- Is [POption (I (mkVersion [0]) InRepo) Nothing]
+-- >>> tryVs [i0, i1]
+-- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-0-inplace"))) Nothing,POption (I (mkVersion [1]) (Inst (UnitId "foo-1-inplace"))) Nothing]
+-- >>> tryVs [i0, v1]
+-- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-0-inplace"))) Nothing,POption (I (mkVersion [1]) InRepo) Nothing]
+-- >>> tryVs [v0, i1]
+-- Is [POption (I (mkVersion [0]) InRepo) Nothing,POption (I (mkVersion [1]) (Inst (UnitId "foo-1-inplace"))) Nothing]
+-- >>> tryVs [i0]
+-- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-0-inplace"))) Nothing]
+-- >>> tryVs []
+-- Is []
+tryVs :: [POption] -> IsOrVs
+tryVs xs@[] = Is xs
+tryVs xs@[_] = Is xs
+tryVs xs
+    | any (\(POption (instI -> b0) (isJust -> b1)) -> b0 || b1) xs = Is xs
+    | otherwise =
+      let (vs, is) = L.partition ((== InRepo) . snd) [(v, l) | POption i _ <- xs, let I v l = i]
+      in if null is then Vs (fst `map` vs) else Is xs
+
+-- | Shows a list of versions in a human-friendly way, abbreviated. Shows a list
+-- of instances in full.
+-- >>> showIsOrVs fooQPN $ tryVs [v0, v1]
+-- "foo; 1, 0"
+-- >>> showIsOrVs fooQPN $ tryVs [v0]
+-- "foo-0"
+-- >>> showIsOrVs fooQPN $ tryVs [i0, i1]
+-- "foo-1/installed-inplace, foo-0/installed-inplace"
+-- >>> showIsOrVs fooQPN $ tryVs [i0, v1]
+-- "foo-1, foo-0/installed-inplace"
+-- >>> showIsOrVs fooQPN $ tryVs [v0, i1]
+-- "foo-1/installed-inplace, foo-0"
+-- >>> showIsOrVs fooQPN $ tryVs []
+-- "unexpected empty list of versions"
+showIsOrVs :: QPN -> IsOrVs -> String
+showIsOrVs _ (Is []) = "unexpected empty list of versions"
+showIsOrVs q (Is [x]) = showOption q x
```

Comment:
I've done this and let intercalate check for singleton lists (it does this via calling `intersperse` that uses recursion with `[]` as a first match).

```haskell
build   :: forall a. (forall b. (a -> b -> b) -> b -> b) -> [a]
{-# INLINE [1] build #-}
        -- The INLINE is important, even though build is tiny,
        -- because it prevents [] getting inlined in the version that
        -- appears in the interface file.  If [] *is* inlined, it
        -- won't match with [] appearing in rules in an importing module.
        --
        -- The "1" says to inline in phase 1

build g = g (:) []

concat :: Foldable t => t [a] -> [a]
concat xs = build (\c n -> foldr (\x y -> foldr c y x) n xs)

intersperse :: a -> [a] -> [a]
intersperse _   []      = []
intersperse sep (x:xs)  = x : prependToAll sep xs

intercalate :: [a] -> [[a]] -> [a]
intercalate xs xss = concat (intersperse xs xss)
```